### PR TITLE
silo-3527: detect asset loss

### PR DESCRIPTION
**Pull request by 0xNeoRacer**
Fixes: silo-3527

Total fixes for silo-3527
- [x] decimals offset for idle vault
- [x] **detect asset loss - this PR**
- [ ] adjust CAP flow

### CODE REVIEW

File: silo-vaults/contracts/SiloVault.sol [hunk 1]
```diff
@@ -1006,4 +1009,12 @@ contract SiloVault is ERC4626, ERC20Permit, Ownable2Step, Multicall, ISiloVaultS
     function _ERC20BalanceOf(address _token, address _account) internal view returns (uint256 balance) {
         balance = IERC20(_token).balanceOf(_account);
     }
+
+    function _assetLossCheck(address _user, uint256 _shares, uint256 _expectedAssets) internal view {
+        uint256 preview = previewRedeem(_shares);
+        uint256 loss = _expectedAssets - preview;
```
Comment (IhorSF) [inline silo-vaults/contracts/SiloVault.sol:1015]: Idk, if this is possible, but I would handle a case when we received more.
```suggestion
        if (preview >= _expectedAssets) return;

        uint256 loss = _expectedAssets - preview;
```

Comment (0xNeoRacer) [reply silo-vaults/contracts/SiloVault.sol:1015]: if you deposit and get more immediately, then:
- we have a bug. 
- or we using some custom vault that can distribute rewards immediately on hook... hmm it might be possible

ok I will support it

Comment (0xNeoRacer) [reply silo-vaults/contracts/SiloVault.sol:1017]: done

File: silo-vaults/contracts/SiloVault.sol [hunk 2]
```diff
@@ -1006,4 +1009,14 @@ contract SiloVault is ERC4626, ERC20Permit, Ownable2Step, Multicall, ISiloVaultS
     function _ERC20BalanceOf(address _token, address _account) internal view returns (uint256 balance) {
         balance = IERC20(_token).balanceOf(_account);
     }
+
+    function _assetLossCheck(address _user, uint256 _shares, uint256 _expectedAssets) internal view {
+        uint256 preview = previewRedeem(_shares);
+        if (preview > _expectedAssets) return;
```
Comment (IhorSF) [inline silo-vaults/contracts/SiloVault.sol:1015]: If it is the same no need to do further checks
```suggestion
        if (preview >= _expectedAssets) return;
```

File: silo-vaults/contracts/SiloVault.sol [hunk 3]
```diff
@@ -1006,4 +1009,14 @@ contract SiloVault is ERC4626, ERC20Permit, Ownable2Step, Multicall, ISiloVaultS
...
+
+        uint256 loss = _expectedAssets - preview;
+        uint256 arbitraryLossThreshold = 10;
+
+        require(loss < arbitraryLossThreshold, ErrorsLib.AssetLoss(loss));
```
Comment (IhorSF) [inline silo-vaults/contracts/SiloVault.sol:1020]: Why not a constant value?
```suggestion
        require(loss < ARBITRARY_LOSS_TRESHOLD, ErrorsLib.AssetLoss(loss));
```

Comment (IhorSF) [reply silo-vaults/contracts/SiloVault.sol:1022]: The size is critical for us. Let's keep it as it is.

File: silo-vaults/contracts/SiloVault.sol [hunk 4]
```diff
@@ -1006,4 +1009,16 @@ contract SiloVault is ERC4626, ERC20Permit, Ownable2Step, Multicall, ISiloVaultS
     function _ERC20BalanceOf(address _token, address _account) internal view returns (uint256 balance) {
         balance = IERC20(_token).balanceOf(_account);
     }
+
+    function _assetLossCheck(address _user, uint256 _shares, uint256 _expectedAssets) internal view {
```
Comment (IhorSF) [inline silo-vaults/contracts/SiloVault.sol:1013]: I don't see that we use a `_user`
```suggestion
    function _assetLossCheck(uint256 _shares, uint256 _expectedAssets) internal view {
```

Comment (0xNeoRacer) [reply silo-vaults/contracts/SiloVault.sol]: done

File: silo-vaults/contracts/SiloVault.sol [hunk 5]
```diff
@@ -605,6 +607,7 @@ contract SiloVault is ERC4626, ERC20Permit, Ownable2Step, Multicall, ISiloVaultS
         assets = _convertToAssetsWithTotals(_shares, totalSupply(), newTotalAssets, Math.Rounding.Ceil);
 
         _deposit(_msgSender(), _receiver, assets, _shares);
+        _assetLossCheck(_receiver, _shares, assets);
```
[the remark this answers is not in the document]
Comment (0xNeoRacer) [reply silo-vaults/contracts/SiloVault.sol]: done

**Review by IhorSF — APPROVED** [review]

File: silo-vaults/test/foundry/IdleVaultTest.sol [hunk 6]
```diff
@@ -146,9 +150,13 @@ contract IdleVaultTest is IntegrationTest {
...
-        assertGe(
-            supplierWithdraw,
-            supplierDeposit < 18 ? 0 : supplierDeposit - 18,
+        uint256 diff = supplierDeposit - supplierWithdraw;
+        uint256 supplierLostPercent = diff * 1e18 / supplierDeposit;
+        emit log_named_uint("supplierLostPercent", supplierLostPercent);
+
+        assertLe(
+            diff,
+            19, // NOTICE: 19 wei can be 50% loss for dust deposits
```
Comment (0xNeoRacer) [inline silo-vaults/test/foundry/IdleVaultTest.sol:159]: hmm why I need 19 if treshold is 10...

Comment (0xNeoRacer) [reply silo-vaults/test/foundry/IdleVaultTest.sol:200]: ah, this is teso on idle vault, not SiloVaults. all good

**Review by IhorSF — APPROVED** [review]
